### PR TITLE
Feature/restore cache

### PR DESCRIPTION
**Pull request by seesharper**
Adds support for a "restore cache" which means that we only perform the underlying `dotnet restore` if the dependencies has changed. This is for instance very useful when editing code in the IDE without changing the dependencies of the script.  Invoking the external `dotnet restore` takes well over 1 second and the user should notice this improvement when for instance starting the debugger in Vs Code.

Note that we can only do this in a safe manner if the script(s) references "pinned" Nuget packages meaning packages with an exact version number. The reason for this is quite simple. If a new version of a package gets released after the cache has been created, we would otherwise run the risk of executing the script with a stale dependency graph.

In addition to all this, the classes related to the project system has been cleaned up a lot and we now introduce value objects for things like `PackageVersion`, `PackageId` and so on. This isolates the semantics of these values such as a `PackageId` being case insensitive. It also cleans up this  whole `Equals` story related to package references.

### CODE REVIEW

File: src/Dotnet.Script.DependencyModel/Runtime/RuntimeDependencyResolver.cs [hunk 1]
```diff
@@ -48,7 +48,7 @@ public RuntimeDependencyResolver(LogFactory logFactory)
         private static IRestorer[] CreateRestorers(LogFactory logFactory)
         {
             var commandRunner = new CommandRunner(logFactory);
-            return new IRestorer[] { new DotnetRestorer(commandRunner, logFactory) };
+            return new IRestorer[] {new ProfiledRestorer(new CachedRestorer(new DotnetRestorer(commandRunner, logFactory),logFactory),logFactory), new NuGetRestorer(commandRunner, logFactory)};
```
[the remark this answers is not in the document]
Comment (seesharper) [reply src/Dotnet.Script.DependencyModel/Runtime/RuntimeDependencyResolver.cs]: Actually, the `NuGetRestorer` is a blast from the past and it was ment as a fallback if we did not have the dotnet cli available. We can clean this up even more by simply removing the `NuGetRestorer` 👍

[the remark this answers is not in the document]
Comment (seesharper) [reply src/Dotnet.Script.DependencyModel/Runtime/RuntimeDependencyResolver.cs]: Makes sense. I'll update to respect the flag 👍

**Comment by seesharper** [comment]
On a side note, the `Program.cs` file needs some major cleanup. There are far to many parameters for a lot of the methods and also some duplicate code. We'll do that in a separate PR 👍

**Comment by seesharper** [comment]
Note that the caching has been disabled for compilation dependencies requested from OmniSharp. There is currently no way to specify something similar to the `--nocache` flag from OmniSharp. We might look into that in more detail when we get a way to hook into OmniSharp so that OmniSharp can discover new NuGet packages. Today we need to restart OmniSharp to discover new packages added to the script(s).

**Comment by filipw** [comment]
thanks 🍻

**Pull request merged**